---
title: "Analysis of all TCGA"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r librarys}
library(FirebrowseR)
library(BiocInstaller)
library(biomaRt)
library(knitr)
library(rlist)
library(edgeR)
library(statmod)
```

```{r settings}
## this can be knitted using rmarkdown::render('tfm1.Rmd')
TASK <- "grosell"
HOME <- '~'
WD <-  file.path(HOME, TASK) 
dir.create(WD, 'data')

opts_chunk$set(fig.width = 12,
               fig.height = 12,
               fig.path = file.path(WD, TASK, 'RmdFigs'),
               output.dir = WD,
               root.dir = WD,
               cache = TRUE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               message = TRUE)

```
## FIMO
Preparation of FIMO list:

```{bash, eval = FALSE}
mkdir utils

cd $_
# retrieving transfac+fimo
wget https://noble.gs.washington.edu/custom-tracks/fimo.hg19.transfac-0.1.bb

# getting the exec for linux 64 bits

rsync -aP rsync://hgdownload.soe.ucsc.edu/genome/admin/exe/linux.x86_64/bigBedToBed .

# making it runnable

chmod a+x bigBedToBed

# converting to bed

./bigBedToBed fimo.hg19.transfac-0.1.bb test.bed

# checking how a bedfile looks like

head test.bed

# counting the TFBS by TF (head)

cut -f4 test.bed | sort | uniq -c | head

cut -f4 test.bed | sort | uniq -c | sed 's/^\s*//' > fimo_transfac.txt
mv fimo_transfac.txt ../data
```

```{r}
fimo_transfac<- unique(read.table(file.path('data', 'fimo_transfac.txt'),
                                header = FALSE,
                                stringsAsFactors = FALSE))
colnames(fimo_transfac)<- c("Number", "tf")
```

##JASPAR
Preparation of JASPAR lists:

```{r jaspar}
cmd <- 'wget https://noble.gs.washington.edu/custom-tracks/fimo.hg19.jaspar_core-0.1.bb;
utils/bigBedToBed fimo.hg19.jaspar_core-0.1.bb jaspar_hg19.bed'
system(cmd)
file.rename('jaspar_hg19.bed', file.path('data', 'jaspar_hg19.bed'))
file.remove('fimo.hg19.jaspar_core-0.1.bb')

jaspar_ids <- unique(read.table(file.path('data', 'jaspar_hg19.bed'),
                                header = FALSE,
                                stringsAsFactors = FALSE)$V4)

length(jaspar_ids)
```
Once we have our TF list we need to transfer the TRANSFAC codes to gene_ID of HGNC symbols, in order to do that we use the package biomaRt from bioconductor (first installation will take some time):

```{r martconfig}
# To show which marts are available
listMarts()

# You need the SNP mart
mart <- useMart("ensembl")

# Find homo sapiens
listDatasets(mart)

# This will be the dataset we want to use
dataset <- useDataset("hsapiens_gene_ensembl", mart=mart)

# Show available filters
listFilters(dataset)

# Now list all available attributes
listAttributes(dataset)

# To get the ensembl gene id belonging to the SNPs
hgnc.factor.fimo.list<-getBM(attributes=c('hgnc_symbol'), 
      filters = 'external_gene_name', 
      values = fimo_transfac$tf, 
      mart = dataset)

hgnc.factor.fimo.list<-as.list(hgnc.factor.fimo.list)
(hgnc.factor.fimo.list)

hgnc.factor.jaspar.list<-getBM(attributes=c('hgnc_symbol'), 
      filters = 'external_gene_name', 
      values = jaspar_ids, 
      mart = dataset)

hgnc.factor.jaspar.list<-as.list(hgnc.factor.jaspar.list)
(hgnc.factor.jaspar.list)
```

The preparation of this lists is on the List.rmd
```{r TF lists}

TF.list<-sort(unique(c(hgnc.factor.fimo.list$hgnc_symbol, hgnc.factor.jaspar.list$hgnc_symbol)))
```


##BREAST:
Donwloading the dataset:
```{r brca_download}
download.file('http://gdac.broadinstitute.org/runs/stddata__2016_01_28/data/BRCA/20160128/gdac.broadinstitute.org_BRCA.Merge_rnaseqv2__illuminahiseq_rnaseqv2__unc_edu__Level_3__RSEM_genes_normalized__data.Level_3.2016012800.0.0.tar.gz',  destfile = file.path(WD, 'brca_rsem.tar.gz'))

brca.fn <- grep('data.txt', untar(file.path(WD, 'brca_rsem.tar.gz'), list = TRUE), value = TRUE)
untar(file.path(WD, 'brca_rsem.tar.gz'), files = brca.fn)

header.brca <- read.table(file = brca.fn, sep = '\t', header = FALSE, nrow = 1,
                     stringsAsFactors = FALSE)

brca <- read.table(file = brca.fn, sep = '\t', header = TRUE, skip = 2)
colnames(brca) <- header.brca[1,]

rownames(brca) <- brca[,1]
brca <- brca[,-1]
rm(header.brca)
```

```{r}
brca.TF.list<-brca[unique(grep(paste(c(TF.list),collapse = "|"), rownames(brca))),]
```

```{r}
brca.TF.list<-round(brca.TF.list)
xy.brca<-DGEList(brca.TF.list)
xo.brca <- order(rowSums(xy.brca$counts), decreasing=TRUE)
xy.brca <- xy.brca[xo.brca,]
xd.brca <- duplicated(xy.brca$counts)
xy.brca <- xy.brca[!xd.brca,]
nrow(xy.brca)
```

```{r}
xy.brca.norm<-calcNormFactors(xy.brca)
xy.brca.norm$samples
```

No es util este plot
```{r eval=FALSE, include=FALSE}
plotMDS(xy.brca.norm)
```

Para diferenciar, acorde a los barcode operaremos:
```{r}
samples.names.brca<-sapply(strsplit(names(brca.TF.list), split = '-', fixed = TRUE), function(x) return(x[]))
table(samples.names.brca)
```

Design matrix:
```{r}
data.frame(sample=(samples.names.brca))
Normal.brca.A<-factor(c("11A" == samples.names.brca[4,]))
Normal.brca.B<-factor(c("11B" == samples.names.brca[4,]))
tumor.brca.A<-factor(c("01A" == samples.names.brca[4,]))
tumor.brca.B<-factor(c("01B" == samples.names.brca[4,]))
brca.matrix<-model.matrix(~Normal.brca.A+tumor.brca.A+Normal.brca.B+tumor.brca.B)
rownames(brca.matrix)<-colnames(xy.brca.norm)
brca.matrix
```

```{r}
xy.brca.disp<-estimateDisp(xy.brca.norm, brca.matrix, robust = FALSE)
#No hay cambios entre true o false (supongo que al no suministrar una matriz cambia la cosa)
xy.brca.disp$common.dispersion
```

```{r}
plotBCV(xy.brca.disp)
```

Differential expression:
```{r}
brca.fit<-glmFit(xy.brca.disp, brca.matrix)
brca.lrt<-glmLRT(brca.fit)
topTags(brca.lrt)
```

```{r}
o.brca <- order(brca.lrt$table$PValue)
cpm(xy.brca.disp)[o.brca[1:10],]
```

```{r}
summary(decideTests(brca.lrt))
```

```{r}
plotMD(brca.lrt)
abline(h=c(-1, 1), col="blue")
```

##COAD:
```{r coad_download}
download.file('http://gdac.broadinstitute.org/runs/stddata__2016_01_28/data/COAD/20160128/gdac.broadinstitute.org_COAD.Merge_rnaseqv2__illuminahiseq_rnaseqv2__unc_edu__Level_3__RSEM_genes_normalized__data.Level_3.2016012800.0.0.tar.gz',  destfile = file.path(WD, 'coad_rsem.tar.gz'))

coad.fn <- grep('data.txt', untar(file.path(WD, 'coad_rsem.tar.gz'), list = TRUE), value = TRUE)
untar(file.path(WD, 'coad_rsem.tar.gz'), files = coad.fn)

header.coad <- read.table(file = coad.fn, sep = '\t', header = FALSE, nrow = 1,
                     stringsAsFactors = FALSE)

coad <- read.table(file = coad.fn, sep = '\t', header = TRUE, skip = 2)
colnames(coad) <- header.coad[1,]

rownames(coad) <- coad[,1]
coad <- coad[,-1]
rm(header.coad)
```

```{r}
coad.TF.list<-round(coad.TF.list)
xy.coad<-DGEList(coad.TF.list)
xo.coad <- order(rowSums(xy.coad$counts), decreasing=TRUE)
xy.coad <- xy.coad[xo.coad,]
xd.coad <- duplicated(xy.coad$counts)
xy.coad <- xy.coad[!xd.coad,]
nrow(xy.coad)
```

```{r}
xy.coad.norm<-calcNormFactors(xy.coad)
xy.coad.norm$samples
```

```{r}
plotMDS(xy.coad.norm)
```

Para diferenciar, acorde a los barcode operaremos:
```{r}
samples.names.coad<-sapply(strsplit(names(coad.TF.list), split = '-', fixed = TRUE), function(x) return(x[]))
table(samples.names.coad)
```

Design matrix:
```{r}
data.frame(sample=(samples.names.coad))
Normal.coad<-factor(c("11A"== samples.names.coad[4,]))
tumor.coad.A<-factor(c("01A"== samples.names.coad[4,]))
tumor.coad.B<-factor(c("01B"== samples.names.coad[4,]))
coad.matrix<-model.matrix(~Normal.coad+tumor.coad.A+tumor.coad.B)
rownames(coad.matrix)<-colnames(xy.coad.norm)
coad.matrix
```

```{r}
xy.coad.disp<-estimateDisp(xy.coad.norm, coad.matrix, robust = FALSE)
#No hay cambios entre true o false (supongo que al no suministrar una matriz cambia la cosa)
xy.coad.disp$common.dispersion
```

```{r}
plotBCV(xy.coad.disp)
```

Differential expression:
```{r}
coad.fit<-glmFit(xy.coad.disp, coad.matrix)
coad.lrt<-glmLRT(coad.fit)
topTags(coad.lrt)
```

```{r}
o.coad <- order(coad.lrt$table$PValue)
cpm(xy.coad.disp)[o.coad[1:10],]
```

```{r}
summary(decideTests(coad.lrt))
```

```{r}
plotMD(coad.lrt)
abline(h=c(-1, 1), col="blue")
```


##LUAD:
Donwloading the dataset:
```{r luad_download}
download.file('http://gdac.broadinstitute.org/runs/stddata__2016_01_28/data/LUAD/20160128/gdac.broadinstitute.org_LUAD.Merge_rnaseqv2__illuminahiseq_rnaseqv2__unc_edu__Level_3__RSEM_genes_normalized__data.Level_3.2016012800.0.0.tar.gz',  destfile = file.path(WD, 'luad_rsem.tar.gz'))

luad.fn <- grep('data.txt', untar(file.path(WD, 'luad_rsem.tar.gz'), list = TRUE), value = TRUE)
untar(file.path(WD, 'luad_rsem.tar.gz'), files = luad.fn)

header.luad <- read.table(file = luad.fn, sep = '\t', header = FALSE, nrow = 1,
                     stringsAsFactors = FALSE)

luad <- read.table(file = luad.fn, sep = '\t', header = TRUE, skip = 2)
colnames(luad) <- header.luad[1,]

rownames(luad) <- luad[,1]
luad <- luad[,-1]
rm(header.luad)
```

```{r}
luad.TF.list<-luad[unique(grep(paste(c(TF.list),collapse = "|"), rownames(luad))),]
```


```{r}
luad.TF.list<-round(luad.TF.list)
xy.luad<-DGEList(luad.TF.list)
xo.luad <- order(rowSums(xy.luad$counts), decreasing=TRUE)
xy.luad <- xy.luad[xo.luad,]
xd.luad <- duplicated(xy.luad$counts)
xy.luad <- xy.luad[!xd.luad,]
nrow(xy.luad)
```

```{r}
xy.luad.norm<-calcNormFactors(xy.luad)
xy.luad.norm$samples
```

Este plot no es util
```{r eval=FALSE, include=FALSE}
plotMDS(xy.luad.norm)
```

Para diferenciar, acorde a los barcode operaremos:
```{r}
samples.names.luad<-sapply(strsplit(names(luad.TF.list), split = '-', fixed = TRUE), function(x) return(x[]))
table(samples.names.luad)
```

Design matrix:
```{r}
data.frame(sample=(samples.names.luad))
Normal.luad.A<-factor(c("11A"== samples.names.luad[4,]))
Normal.luad.B<-factor(c("11B"== samples.names.luad[4,]))
tumor.luad.A<-factor(c("01A"== samples.names.luad[4,]))
tumor.luad.B<-factor(c("01B"== samples.names.luad[4,]))
tumor.luad.2A<-factor(c("02A"== samples.names.luad[4,]))
luad.matrix<-model.matrix(~Normal.luad.A+tumor.luad.A+Normal.luad.B+tumor.luad.B)
rownames(luad.matrix)<-colnames(xy.luad.norm)
luad.matrix
```

Da error si admitimos el tumor 02A en la matriz, no asi como en el resto, son dos casos no se si se puede omitir
O por el contrario hay que hacer algo mas.
```{r}
xy.luad.disp<-estimateDisp(xy.luad.norm, luad.matrix, robust = FALSE)
#No hay cambios entre true o false (supongo que al no suministrar una matriz cambia la cosa)
xy.luad.disp$common.dispersion
```

```{r}
plotBCV(xy.luad.disp)
```

Differential expression:
```{r}
luad.fit<-glmFit(xy.luad.disp, luad.matrix)
luad.lrt<-glmLRT(luad.fit)
topTags(luad.lrt)
```

```{r}
o.luad <- order(luad.lrt$table$PValue)
cpm(xy.luad.disp)[o.luad[1:10],]
```

```{r}
summary(decideTests(luad.lrt))
```

```{r}
plotMD(luad.lrt)
abline(h=c(-1, 1), col="blue")
```


##PRAD:
Donwloading the dataset:
```{r prad_download}
download.file('http://gdac.broadinstitute.org/runs/stddata__2016_01_28/data/PRAD/20160128/gdac.broadinstitute.org_PRAD.Merge_rnaseqv2__illuminahiseq_rnaseqv2__unc_edu__Level_3__RSEM_genes_normalized__data.Level_3.2016012800.0.0.tar.gz',  destfile = file.path(WD, 'prad_rsem.tar.gz'))

prad.fn <- grep('data.txt', untar(file.path(WD, 'prad_rsem.tar.gz'), list = TRUE), value = TRUE)
untar(file.path(WD, 'prad_rsem.tar.gz'), files = prad.fn)

header.prad <- read.table(file = prad.fn, sep = '\t', header = FALSE, nrow = 1,
                     stringsAsFactors = FALSE)

prad <- read.table(file = prad.fn, sep = '\t', header = TRUE, skip = 2)
colnames(prad) <- header.prad[1,]

rownames(prad) <- prad[,1]
prad <- prad[,-1]
rm(header.prad)
```

```{r}
prad.TF.list<-prad[unique(grep(paste(c(TF.list),collapse = "|"), rownames(prad))),]
```

```{r}
prad.TF.list<-round(prad.TF.list)
xy.prad<-DGEList(prad.TF.list)
xo.prad <- order(rowSums(xy.prad$counts), decreasing=TRUE)
xy.prad <- xy.prad[xo.prad,]
xd.prad <- duplicated(xy.prad$counts)
xy.prad <- xy.prad[!xd.prad,]
nrow(xy.prad)
```

```{r}
xy.prad.norm<-calcNormFactors(xy.prad)
xy.prad.norm$samples
```

```{r}
plotMDS(xy.prad.norm)
```

Para diferenciar, acorde a los barcode operaremos:
```{r}
samples.names.prad<-sapply(strsplit(names(prad.TF.list), split = '-', fixed = TRUE), function(x) return(x[]))
table(samples.names.prad)
```

Design matrix:
```{r}
data.frame(sample=(samples.names.prad))
Normal.prad.A<-factor(c("11A"== samples.names.prad[4,]))
Normal.prad.B<-factor(c("11B"== samples.names.prad[4,]))
tumor.prad.A<-factor(c("01A"== samples.names.prad[4,]))
tumor.prad.B<-factor(c("01B"== samples.names.prad[4,]))
prad.matrix<-model.matrix(~Normal.prad.A+tumor.prad.A+Normal.prad.B+tumor.prad.B)
rownames(prad.matrix)<-colnames(xy.prad.norm)
prad.matrix
```

```{r}
xy.prad.disp<-estimateDisp(xy.prad.norm, prad.matrix, robust = TRUE)
xy.prad.disp$common.dispersion
```

```{r}
plotBCV(xy.prad.disp)
```

Differential expression:
```{r}
prad.fit<-glmFit(xy.prad.disp, prad.matrix)
prad.lrt<-glmLRT(prad.fit)
topTags(prad.lrt)
```

```{r}
o.prad <- order(prad.lrt$table$PValue)
cpm(xy.prad.disp)[o.prad[1:10],]
```

```{r}
summary(decideTests(prad.lrt))
```

```{r}
plotMD(prad.lrt)
abline(h=c(-1, 1), col="blue")
```