---
title: "COAD diff expression matched N and T proposal"
output:
  html_document:
    keep_md: false
    toc: true
    toc_float: true
    toc_depth: 4

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup

```{r libraries}
## library(FirebrowseR)
## library(BiocInstaller)
library(biomaRt)
library(knitr)
## library(rlist)
library(edgeR)
## library(statmod)
library(org.Hs.eg.db)
library(RSQLite) #
```

```{r settings}

opts_chunk$set(fig.width = 12,
               fig.height = 12,
               cache = TRUE,
               include = TRUE,
               cache.lazy = FALSE,
               warning = TRUE,
               message = TRUE)

```
# TFBS data retrieval

For transfac + fimo scans

```{bash, eval = TRUE}
mkdir -p utils 

cd $_
# retrieving transfac+fimo
wget -q https://noble.gs.washington.edu/custom-tracks/fimo.hg19.transfac-0.1.bb

# getting the exec for linux 64 bits

rsync -aP rsync://hgdownload.soe.ucsc.edu/genome/admin/exe/linux.x86_64/bigBedToBed .

# making it runnable

chmod a+x bigBedToBed

# converting to bed

./bigBedToBed fimo.hg19.transfac-0.1.bb test.bed

# checking how a bedfile looks like

head test.bed

cut -f4 test.bed | sort | uniq -c | sed 's/^\s*//' > fimo_transfac.txt
mv fimo_transfac.txt ../data

```

```{r}
fimo_transfac <- unique(read.table(file.path('data', 'fimo_transfac.txt'),
                                header = FALSE,
                                stringsAsFactors = FALSE))
colnames(fimo_transfac)<- c("Number", "tf")

```


# Mapping

## Symbols

We would like to transform TRANSFAC names to HUGO symbols (we'll check differential expression of these using TCGA RSEM data)

```{r mapping}

con <- org.Hs.eg_dbconn()

stmt <- 'SELECT * FROM alias, gene_info WHERE alias._id == gene_info._id;'
# execute the query on the database
aliases <- dbGetQuery(con, stmt)

candidates <- aliases[aliases$alias_symbol %in% toupper(fimo_transfac$tf),]
candidates2 <- aliases[aliases$symbol %in% toupper(fimo_transfac$tf),]

valid_symbols <- unique(candidates$symbol, candidates2$symbol)
```

## Entrez IDs

Gene symbols to Entrez gene ids (TCGA RSEMs also provide these identifiers)

```{r biomart_entrez}

mart <- useMart("ENSEMBL_MART_ENSEMBL", host="grch37.ensembl.org",
                path="/biomart/martservice", dataset="hsapiens_gene_ensembl")


entrez_tf <- getBM(attributes=c('hgnc_symbol', 
                                'entrezgene'), 
                   filters = 'external_gene_name', 
                   values = valid_symbols,
                   mart = mart)

```

# RSEM download for COAD

COAD normalized RSEM expression values download from GDAC

```{r coad_download}

## using firebrowse instead?
download.file('http://gdac.broadinstitute.org/runs/stddata__2016_01_28/data/COAD/20160128/gdac.broadinstitute.org_COAD.Merge_rnaseqv2__illuminahiseq_rnaseqv2__unc_edu__Level_3__RSEM_genes_normalized__data.Level_3.2016012800.0.0.tar.gz',  destfile = file.path('data', 'coad_rsem.tar.gz'))

coad.fn <- grep('data.txt', untar(file.path('data', 'coad_rsem.tar.gz'), list = TRUE), value = TRUE)
untar(file.path('data', 'coad_rsem.tar.gz'), files = coad.fn)

header.coad <- read.table(file = coad.fn, sep = '\t', header = FALSE, nrow = 1,
                     stringsAsFactors = FALSE)

coad <- read.table(file = coad.fn, sep = '\t', header = TRUE, skip = 2)
colnames(coad) <- header.coad[1,]

rownames(coad) <- coad[,1]
coad <- coad[,-1]
rm(header.coad)

```

Getting a dictionary of TF with their TCGA id, Entrez id and standard gene symbol for which we know the TRANSFAC/FIMO motif.


```{r}
mapping <- data.frame(tcga = rownames(coad),
                      entrez = sapply(strsplit(rownames(coad), '|', fixed = TRUE),
                                      function(x) return(x[2])),
                      symbol = sapply(strsplit(rownames(coad), '|', fixed = TRUE),
                                      function(x) return(x[1])))

tf <- merge(mapping, entrez_tf, by.x = 'entrez', by.y = 'entrezgene')

head(tf)

## so tf$tcga will contain the TCGA identifiers (rownames of COAD)
dim(coad[tf$tcga,])

```

# Diff expression

Getting sort of counts from RSEMs (this is not correct because we are just rounding RSEMs!) and generating a DGE object only for those samples that come from the same patient, because we will compare tumors against matched normals.

```{r}

coad <- round(coad) ## this is dirty, they are RSEMs and not counts!!!!!!!

coldata <- data.frame(barcode = colnames(coad),
                      sample = strtrim(sapply(strsplit(colnames(coad), '-', fixed = TRUE),
                                              function(x) return(x[4])), 2),
                      participant = sapply(strsplit(colnames(coad), '-', fixed = TRUE),
                                              function(x) return(x[3]))
                      )
table(coldata$sample)
## only normals and primary tumors
coldata <- coldata[coldata$sample %in% c('01', '11'),]

## only matched normal and tumor from the same participant
selected <- coldata$participant[duplicated(coldata$participant)]
coldata <- coldata[coldata$participant %in% selected,]
coad <- coad[,coldata$barcode]

coad_dge <- DGEList(coad, samples = coldata)

```

Norm factors

```{r}
coad_dge <- calcNormFactors(coad_dge)

```
Do tumors match tumors and normals match normals? Or do samples from the same participant cluster together? What does this mean (biologically)?

```{r}
plotMDS(coad_dge)
```

Differential expression analysis goes next.
